pageserver/src/tenant.rs

```diff
@@ -353,6 +358,468 @@ impl Drop for TimelineUninitMark {
+                remote_storage.clone(),
+            )
+            .instrument(info_span!("load_remote_timeline", timeline = %timeline_id))
+            .await?
+        }
+
+        // FIXME: Check if the state has changed to Stopping while we were downloading stuff
+        // We're ready for business.
+        // Change to active state under the hood spawns background loops
+        // The loops will shut themselves down when they notice that the tenant is inactive.
+        self.set_state(TenantState::Active {
+            background_jobs_running: true,
+        });
+
+        info!("Done");
+
+        Ok(())
+    }
+
+    async fn load_remote_timeline(
+        &self,
+        timeline_id: TimelineId,
+        index_part: &IndexPart,
+        remote_storage: GenericRemoteStorage,
+    ) -> anyhow::Result<()> {
+        info!("downloading index file for timeline {}", timeline_id);
+        tokio::fs::create_dir_all(self.conf.timeline_path(&timeline_id, &self.tenant_id))
+            .await
+            .context("Failed to create new timeline directory")?;
+
+        let remote_metadata = index_part.parse_metadata().with_context(|| {
+            format!(
+                "Failed to parse metadata file from remote storage for tenant {}",
+                self.tenant_id
+            )
+        })?;
+
+        let remote_client =
+            create_remote_timeline_client(remote_storage, self.conf, self.tenant_id, timeline_id)?;
+
+        let ancestor = if let Some(ancestor_id) = remote_metadata.ancestor_timeline() {
+            let timelines = self.timelines.lock().unwrap();
+            Some(Arc::clone(timelines.get(&ancestor_id).ok_or_else(
+                || {
+                    anyhow::anyhow!(
+                        "cannot find ancestor timeline {ancestor_id} for timeline {timeline_id}"
+                    )
+                },
+            )?))
+        } else {
+            None
+        };
+
+        self.setup_timeline(
+            timeline_id,
+            Some(remote_client),
+            Some(index_part),
+            remote_metadata,
+            ancestor,
+            true,
+        )
+        .await
+    }
+
+    ///
+    /// Load a tenant that's available on local disk
+    ///
+    /// This is used at pageserver startup, to rebuild the in-memory
+    /// structures from on-disk state. This is similar to attaching a tenant,
+    /// but the index files already exist on local disk, as well as some layer
+    /// files.
+    ///
+    /// If the loading fails for some reason, the Tenant will go into Broken
+    /// state.
+    ///
+    #[instrument(skip(conf, remote_storage), fields(tenant_id=%tenant_id))]
+    pub fn spawn_load(
+        conf: &'static PageServerConf,
+        tenant_id: TenantId,
+        remote_storage: Option<GenericRemoteStorage>,
+    ) -> anyhow::Result<Arc<Tenant>> {
+        // FIXME: also go into Broken state if this fails
+        let tenant_conf = Self::load_tenant_config(conf, tenant_id)?;
```

Comment:
> I think not.

Potentially operator can come and fix config and then reload the tenant locally. As Heikki said above, deleting something is always scary. We can delete things if we're sure that it is a half initialized tenant. This is the only case I think.

Also leaving it in place allows to inspect it afterwards. If we delete everything it would be quite hard to figure out why it failed

Reply:
Ok, so, let's not delete anything.

I think I added the logic for creating a broken tenant already for the case where the local tenant dir is in attaching state:

```
  let tenant = if conf.tenant_attaching_mark_file_path(&tenant_id).exists() {
        info!("tenant {tenant_id} has attaching mark file, resuming its attach operation");
        if let Some(remote_storage) = remote_storage {
            Tenant::spawn_attach(conf, tenant_id, &remote_storage)?
        } else {
            warn!("tenant {tenant_id} has attaching mark file, but pageserver has no remote storage configured");
            // XXX there should be a constructor to make a broken tenant
            // TODO should we use Tenant::load_tenant_config() here?
            let tenant_conf = TenantConfOpt::default();
            let wal_redo_manager = Arc::new(PostgresRedoManager::new(conf, tenant_id));
            let tenant = Tenant::new(
                TenantState::Broken,
                conf,
                tenant_conf,
                wal_redo_manager,
                tenant_id,
                remote_storage,
            );
            Arc::new(tenant)
        }
    } else {
        info!("tenant {tenant_id} is assumed to be loadable, starting load operation");
        // Start loading the tenant into memory. It will initially be in Loading state.
        Tenant::spawn_load(conf, tenant_id, remote_storage)?
    };
```

I'd go ahead and generalize this so that when either `Tenant::spawn_load` or `spawn_attach` fail, we create the broken tenant. Ok?

Reply:
Resolving in https://github.com/neondatabase/neon/pull/2932